File: core/src/main/java/com/nfthub/core/entity/Category.java
package com.nfthub.core.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.*;

@Entity
@Table
@Getter
@NoArgsConstructor
public class Category {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @Column(unique = true, nullable = false)
    private String name;
    @ManyToOne(fetch = FetchType.LAZY)
    private CategoryGroup categoryGroup;

    public Category(String name) {
        this.name = name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setCategoryGroup(CategoryGroup categoryGroup) {
        if (this.categoryGroup != null) {
            this.categoryGroup.getCategories().remove(this);
        }
        this.categoryGroup = categoryGroup;
        categoryGroup.getCategories().add(this);
    }
}

File: src/main/kotlin/com/nfthub/api/entity/Category.kt
package com.nfthub.api.entity

import com.nfthub.api.EMPTY_STRING
import javax.persistence.*

@Entity
@Table
class Category(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(nullable = false)
    var id: Long = 0,
    @Column(unique = true, nullable = false)
    var name: String = EMPTY_STRING,
    @ManyToOne(fetch = FetchType.LAZY, optional = true)
    var categoryGroup: CategoryGroup? = null
)

enum class Category_(val s: String) {
    Id("id"),
    Name("name"),
    CategoryGroup("categoryGroup")
}

@Entity
@Table
class CategoryGroup(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(nullable = false)
    var id: Long = 0,
    @Column(unique = true, nullable = false)
    var name: String = EMPTY_STRING,
    @OneToMany(mappedBy = "categoryGroup")
    var categories: List<Category> = emptyList()
)

enum class CategoryGroup_(val s: String) {
    Name("name"),
    Categories("categories")
}